registry/regsrc/terraform_provider.go

```diff
@@ -0,0 +1,58 @@
+package regsrc
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/hashicorp/terraform/svchost"
+)
+
+var (
+	// DefaultProviderNamespace represents the namespace for canonical
+	// HashiCorp-controlled providers.
+	// REVIEWERS: Naming things is hard.
+	// * HashiCorpProviderNameSpace?
+	// * OfficialP...?
+	// * CanonicalP...?
+	DefaultProviderNamespace = "terraform-providers"
+)
+
+// TerraformProvider describes a Terraform Registry Provider source.
+type TerraformProvider struct {
```

Comment:
Since everything in the registry is a Terraform registry is related to Terraform, maybe just `Provider`? (or was this following some existing precedent for redundancy?)

Reply:
That's a good question and I was confused by this initially, too. I copied the pattern (and most of this code directly) out of the terraform-registry repository. Let me dig back into that and come back...

Reply:
Ok, addressing two bits of feedback here. The registry has both  `type Provider` and `type TerraformProvider`. I'm fuzzy on the details but I believe that (in the registry) `Provider` is intended for the UI  and `TerraformProvider` is the response structure specifically for terraform cli's use. Most of these were copied directly from the registry code; anything that I added I kept consistent as best I could. 

This is not to say I can't just rename `TerraformProvider` to `Provider` - if nothing else I'm realizing that I can remove some of the code I'd copied - but I'm learning towards leaving it as-is (the name, not the unnecessary code) for consistency and adding some comments that explain the difference. Preferences?

Reply:
Oh right the registry _does_ have a second meaning of provider: it's a sort of secondary namespace for modules, so that we can have both `hashicorp/consul/aws` and `hashicorp/consul/azurerm`, which are two separate modules that are assumed (by convention) to get a similar result against two different providers.

Given that the _module_ meaning of "provider" is more of a secondary key than a "real" object, I guess ideally we'd use `type Provider` to represent the first-class provider idea (what you're implementing now) and have something like `type ModuleProvider` to represent that extra containing namespace for modules. But I'd also be fine with just running with this as-is for now to avoid getting too drastic with the refactoring, since (as with the other comment) it'll be easier to see how this all fits together and what makes sense in a later phase when more of the pieces are in place.